pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,437 @@
+// * Extracts all ImageStramTags out of the Image
+// * Filters out the ones that are not in use
+// Note: We can not use a predicate because that is directly applied on the source and the source yields ImageStreams, not ImageStreamTags
+// * Creates a reconcile.Request per cluster and ImageStreamTag
+func registryClusterHandlerFactory(buildClusters sets.String, filter objectFilter) handler.EventHandler {
+	return imagestreamtagmapper.New(func(in reconcile.Request) []reconcile.Request {
+		if !filter(in.NamespacedName) {
+			return nil
+		}
+
+		var requests []reconcile.Request
+		// We have to squeeze both the target cluster name and the imageStreamTag name into a reconcile.Request
+		// Internally, this gets put onto the workqueue as a single string in namespace/name notation and split
+		// later on. This means that we can not use a slash as delimiter for the cluster and the namespace.
+		for _, buildCluster := range buildClusters.List() {
+			name := types.NamespacedName{
+				Namespace: buildCluster + clusterAndNamespaceDelimiter + in.Namespace,
+				Name:      in.Name,
+			}
+			requests = append(requests, reconcile.Request{NamespacedName: name})
+		}
+		return requests
+	})
+}
+
+const clusterAndNamespaceDelimiter = "_"
+
+func decodeRequest(req reconcile.Request) (string, types.NamespacedName, error) {
+	clusterAndNamespace := strings.Split(req.Namespace, "_")
+	if n := len(clusterAndNamespace); n != 2 {
+		return "", types.NamespacedName{}, fmt.Errorf("didn't get two but %d segments when trying to extract cluster and namespace", n)
+	}
+	return clusterAndNamespace[0], types.NamespacedName{Namespace: clusterAndNamespace[1], Name: req.Name}, nil
+}
+
+type reconciler struct {
+	ctx                 context.Context
+	log                 *logrus.Entry
+	registryClients     map[string]ctrlruntimeclient.Client
+	pullSecretGetter    func() []byte
+	forbiddenRegistries sets.String
+	dryRun              bool
+}
+
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(req, log)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req reconcile.Request, log *logrus.Entry) error {
+	cluster, decoded, err := decodeRequest(req)
+	if err != nil {
+		return fmt.Errorf("failed to decode request %s: %w", req, err)
+	}
+
+	// Propagate the cluster field back up
+	*log = *log.WithField("cluster", cluster)
+
+	isTags := map[string]*imagev1.ImageStreamTag{}
+	for ctx, client := range r.registryClients {
+		imageStreamTag := &imagev1.ImageStreamTag{}
+		if err := client.Get(r.ctx, decoded, imageStreamTag); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Debug("Source imageStreamTag not found")
```

Comment:
We don't know what cluster the request came from, so we can say which is the `source`. We should:
* Do nothing if it doesn't exist on both clusters
* Do nothing if its the same on both clusters
* Import if its newer on one cluster or doesn't exist on the other